File: workout-api/src/main/java/tom/projects/monkeylog/controller/ExerciseGroupController.java
package tom.projects.monkeylog.controller;

import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;
import tom.projects.monkeylog.dto.workout.*;
import tom.projects.monkeylog.mapper.WorkoutMapper;
import tom.projects.monkeylog.service.ExerciseGroupService;
import tom.projects.monkeylog.service.ExerciseRowService;

@RestController
@RequiredArgsConstructor
public class ExerciseGroupController {
    private final ExerciseGroupService exerciseGroupService;
    private final ExerciseRowService exerciseRowService;
    private final WorkoutMapper workoutMapper;

    @PostMapping("/exercise_group/{id}/add_row")
    ExerciseGroupResponse addRow(@PathVariable Long id) {
        return workoutMapper.exerciseGroupToExerciseGroupResponse(exerciseGroupService.addRow(id));
    }

    @PostMapping("/exercise_group/{id}/swap_row")
    void swapRow(@RequestBody ExerciseRowSwapRequest exerciseRowSwapRequest, @PathVariable Long id) {
        exerciseGroupService.swapRow(id, exerciseRowSwapRequest);
    }

    @PatchMapping("/exercise_row/{id}")
    ExerciseRowResponse updateRow(@RequestBody ExerciseRowUpdateRequest exerciseRowUpdateRequest,
            @PathVariable Long id) {
        return workoutMapper.exerciseRowToExerciseRowResponse(exerciseRowService.update(exerciseRowUpdateRequest, id));
    }

    @PatchMapping("/exercise_row_field/{exerciseRowFieldId}")
    ExerciseRowFieldResponse updateRowField(@RequestBody ExerciseRowFieldUpdateRequest exerciseRowFieldUpdateRequest,
            @PathVariable Long exerciseRowFieldId) {
        return workoutMapper.exerciseRowFieldToExerciseRowFieldResponse(
                exerciseRowService.updateField(exerciseRowFieldUpdateRequest, exerciseRowFieldId));
    }

    @PostMapping("/workout/{workoutId}/exercise_group")
    void save(@RequestBody ExerciseGroupCreateRequest exerciseGroupCreateRequest, @PathVariable Long workoutId) {
        exerciseGroupService.save(exerciseGroupCreateRequest, workoutId);
    }

    @DeleteMapping("/exercise_row/{id}")
    void destroyRow(@PathVariable Long id) {
        exerciseRowService.delete(id);
    }

    @DeleteMapping("/exercise_group/{id}")
    void destroy(@PathVariable Long id) {
        exerciseGroupService.delete(id);
    }
}

File: backend/src/main/kotlin/tom/com/workout/logbook/controller/ExerciseGroupController.kt
package tom.com.workout.logbook.controller

import io.swagger.v3.oas.annotations.tags.Tag
import org.springdoc.core.annotations.ParameterObject
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import tom.com.workout.logbook.common.dto.IdProjection
import tom.com.workout.logbook.common.dto.Page
import tom.com.workout.logbook.common.mapper.toResponse
import tom.com.workout.logbook.dto.workout.*
import tom.com.workout.logbook.mapper.toResponse
import tom.com.workout.logbook.mapper.toResponseWithWorkout
import tom.com.workout.logbook.service.ExerciseGroupService
import tom.com.workout.logbook.service.ExerciseRowService
import java.util.*

@RestController
@Tag(name = "Workout")
class ExerciseGroupController(
    private val exerciseGroupService: ExerciseGroupService,
    private val exerciseRowService: ExerciseRowService,
) {
    @PostMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}/exercise-rows")
    fun createExerciseRow(
        @PathVariable workoutId: UUID,
        @PathVariable exerciseGroupId: UUID
    ) = exerciseGroupService.addRow(exerciseGroupId).toResponse()

    @PutMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}/exercise-rows/{exerciseRowId}")
    fun updateExerciseRow(
        @PathVariable workoutId: UUID,
        @PathVariable exerciseGroupId: UUID,
        @PathVariable exerciseRowId: UUID,
        @RequestBody exerciseRowUpdateRequest: ExerciseRowUpdateRequest
    ) = exerciseRowService.update(
        exerciseRowUpdateRequest,
        exerciseRowId
    ).toResponse()

    @PostMapping("/workouts/{workoutId}/exercise-groups")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun createExerciseGroup(
        @PathVariable workoutId: UUID,
        @RequestBody exerciseGroupCreateRequest: ExerciseGroupCreateRequest
    ) = exerciseGroupService.save(exerciseGroupCreateRequest, workoutId)

    @DeleteMapping("/workouts/{workoutId}/exercise-group/{exerciseGroupId}/exercise-rows/{exerciseRowId}")
    @ResponseStatus(
        HttpStatus.NO_CONTENT
    )
    fun deleteExerciseRow(
        @PathVariable workoutId: UUID,
        @PathVariable exerciseGroupId: UUID,
        @PathVariable exerciseRowId: UUID
    ) = exerciseRowService.delete(exerciseRowId)

    @DeleteMapping("/workouts/{workoutId}/exercise-groups/{exerciseGroupId}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteExerciseGroup(@PathVariable workoutId: UUID, @PathVariable exerciseGroupId: UUID) =
        exerciseGroupService.delete(exerciseGroupId)

    @GetMapping("/exercises/{exerciseId}/exercise-groups")
    fun getExerciseGroups(
        @PathVariable exerciseId: UUID,
        @ParameterObject pageable: Pageable
    ): Page<ExerciseGroupWithWorkoutResponse> {
        val uuids = exerciseGroupService.getExerciseGroupsByExerciseId(exerciseId, pageable)
        val exerciseGroups =
            exerciseGroupService.getExerciseGroups(uuids.content.map(IdProjection::id))
                .associate { it.id!! to it.toResponseWithWorkout() }

        return uuids.toResponse { exerciseGroups[it.id]!! }
    }
}

